Add a tutorial to load metrics from prometheus server

yongtang opened the pull request:
This PR adds tutorial to load metrics from prometheus server
into tf.data.Dataset, so that it could be used for tf.keras
and model training.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
This is another tutorial to load metrics data from prometheus server. prometheus is almost a common setup in kubernetes, and is widely seem in cloud-native devops/infrastructure. this tutorial could help devops/sre to use tensorflow/tensorflow-io to automate their operations.

yongtang commented:
/cc @MarkDaoust @lamberta  to take a look.

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+      "source": [
+        "## Overview\n",
+        "\n",
+        "This tutorial will show how to load CoreDNS metrics into  `tf.data.Dataset` from a [Prometheus](https://prometheus.io) server, so that the loaded metrics could be represented as a `tf.data.Dataset` and passed to `tf.keras` for training or inference purposes.\n",
```
```suggestion
        "This tutorial loads CoreDNS metrics from a [Prometheus](https://prometheus.io) server into a `tf.data.Dataset`, then uses `tf.keras` for training and inference.\n",
```

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+      "metadata": {
+        "id": "yZmI7l_GykcW",
+        "colab_type": "text"
+      },
```
I'd create another code cell here that imports all the modules below. This way they're all in one place in the setup section.

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+      "source": [
+        "### Install and setup CoreDNS and Prometheus\n",
+        "\n",
+        "For demo purposes, a CoreDNS server locally with port `9053` open to receive DNS queries and port `9153` (defult) open to expose metrics for scraping. A simple Corefile configuration of CoreDNS is as follows and is available to be [downloaded](https://github.com/tensorflow/io/blob/master/docs/tutorials/prometheus/Corefile):\n",
```
```suggestion
        "For demo purposes, a CoreDNS server locally with port `9053` open to receive DNS queries and port `9153` (defult) open to expose metrics for scraping. The following is a basic Corefile configuration for CoreDNS iand is available to [download](https://github.com/tensorflow/io/blob/master/docs/tutorials/prometheus/Corefile):\n",
```

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+        "\n",
+        "# Run `./coredns` as a background process.\n",
+        "# IPython doesn't recognize `&` in inline bash cells.\n",
+        "get_ipython().system_raw('./coredns &')"
```
Use separate code cells for shell commands and python. We have styles that make this clearer when imported into the website.

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+        "\n",
+        "Create a Dataset for CoreDNS metrics that is available from PostgreSQL server, could be done with `tfio.experimental.IODataset.from_prometheus`. At the minimium two arguments are needed. `query` is passed to Prometheus server to select the metrics and `length` is the period we want to load into Dataset.\n",
+        "\n",
+        "You can start with \"coredns_dns_request_count_total\" and \"5\" (secs) to create the Dataset below. Since earlier in the tutorial two DNS queries were sent, it is expected that the metrics for \"coredns_dns_request_count_total\" will be \"2.0\" at the end of the time series:"
```
```suggestion
        "You can start with `\"coredns_dns_request_count_total\"` and `\"5\"` (secs) to create the Dataset below. Since earlier in the tutorial two DNS queries were sent, it is expected that the metrics for `\"coredns_dns_request_count_total\"` will be `\"2.0\"` at the end of the time series:"
```

lamberta commented on docs/tutorials/prometheus.ipynb:
````diff
@@ -0,0 +1,471 @@
+        "}\n",
+        "```\n",
+        "\n",
+        "In the above example, 'coredns' is the job name, 'localhost:9153' is the instance name, and 'coredns_dns_request_count_total' is the metric name. Note that depending on the Prometheus query used, it is possible that multiple jobs/instances/metrics could be returned. This is also the reason why python dict has been used in the structure of the Dataset.\n",
````
```suggestion
        "In the above example, `'coredns'` is the job name, `'localhost:9153'` is the instance name, and `'coredns_dns_request_count_total'` is the metric name. Note that depending on the Prometheus query used, it is possible that multiple jobs/instances/metrics could be returned. This is also the reason why python dict has been used in the structure of the Dataset.\n",
```

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+        "\n",
+        "In the above example, 'coredns' is the job name, 'localhost:9153' is the instance name, and 'coredns_dns_request_count_total' is the metric name. Note that depending on the Prometheus query used, it is possible that multiple jobs/instances/metrics could be returned. This is also the reason why python dict has been used in the structure of the Dataset.\n",
+        "\n",
+        "Take another query \"go_memstats_gc_sys_bytes\" as an example. Since both CoreDNS and Prometheus are written in Golang, \"go_memstats_gc_sys_bytes\" metric is available for both \"coredns\" job and \"prometheus\" job:"
```
```suggestion
        "Take another query `\"go_memstats_gc_sys_bytes\"` as an example. Since both CoreDNS and Prometheus are written in Golang, `\"go_memstats_gc_sys_bytes\"` metric is available for both `\"coredns\"` job and `\"prometheus\"` job:"
```

lamberta commented on docs/tutorials/prometheus.ipynb:
```diff
@@ -0,0 +1,471 @@
+        "from datetime import datetime\n",
+        "import tensorflow_io as tfio\n",
+        "\n",
+        "dataset = tfio.experimental.IODataset.from_prometheus(\n",
```
Getting an error here when run on Colab. Is this expected?

@MarkDaoust Since this notebook is spinning up servers and such, will this work when imported into tensorflow.org? Do we need to save the outputs?

Trace:
```
WARNING:tensorflow:AutoGraph could not transform <function io_prometheus_readable_read at 0x7f5de5f60e18> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Unreachable
WARNING: AutoGraph could not transform <function io_prometheus_readable_read at 0x7f5de5f60e18> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Unreachable
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-6-5ea8e8be172d> in <module>()
      3 
      4 dataset = tfio.experimental.IODataset.from_prometheus(
----> 5     "coredns_dns_request_count_total", 5)
      6 
      7 print("Dataset Spec:\n{}\n".format(dataset.element_spec))

12 frames
/tensorflow-2.1.0/python3.6/tensorflow_core/python/autograph/impl/api.py in wrapper(*args, **kwargs)
    235       except Exception as e:  # pylint:disable=broad-except
    236         if hasattr(e, 'ag_error_metadata'):
--> 237           raise e.ag_error_metadata.to_exception(e)
    238         else:
    239           raise

AssertionError: in converted code:

    /usr/local/lib/python3.6/dist-packages/tensorflow_io/core/python/ops/prometheus_dataset_ops.py:82 f  *
        timestamp, value = golang_ops.io_prometheus_readable_read(
    <string>:254 io_prometheus_readable_read
        
    /tensorflow-2.1.0/python3.6/tensorflow_core/python/framework/op_def_library.py:602 _apply_op_helper
        assert False, "Unreachable"

    AssertionError: Unreachable
```

MarkDaoust replied:
> Since this notebook is spinning up servers and such, will this work when imported into tensorflow.org?

I think this part shouldn't be a problem.

yongtang commented:
Thanks @lamberta  @MarkDaoust for the review. I have updated the PR to place import modules into one section, and split the python and shell code, as was suggested in the review.

I also move the server setup into an early section, that should solve the potential `  AssertionError: Unreachable` issue. The issue is likely caused by server not ready yet (may take a second or two to setup.)

Please take a look and let me know if there are any other issues.
